src/appleseed/foundation/math/sampling/sphericalrectanglesampler.h

```diff
@@ -0,0 +1,164 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#pragma once
+
+// appleseed.foundation headers.
+#include "foundation/math/scalar.h"
+#include "foundation/math/vector.h"
+
+// Standard headers.
+#include <cassert>
+#include <cmath>
+
+namespace foundation
+{
+
+//
+// References:
+//
+//   An Area-Preserving Parametrization for Spherical Rectangles.
+//   https://www.arnoldrenderer.com/research/egsr2013_spherical_rectangle.pdf
+//
+
+template <typename T>
+class SphericalRectangleSampler
+{
+  public:
+    SphericalRectangleSampler(
+        const Vector<T,3>& origin,
+        const Vector<T,3>& x_axis,
+        const Vector<T,3>& y_axis,
+        const Vector<T,3>& normal,
+        const Vector<T,3>& o)
+      : m_x(x_axis)
+      , m_y(y_axis)
+      , m_z(normal)
+      , m_o(o)
+    {
+        const Vector<T,3> d = origin - o;
+
+        const T width = norm(x_axis);
+        const T height = norm(y_axis);
+
+        const T x0 = dot(d, x_axis);
+        const T x1 = x0 + width;
+
+        m_y0 = dot(d, y_axis);
+        m_y1 = m_y0 + height;
+        m_z0 = dot(d, normal);
+
+        // z flip
+        if (m_z0 > T(0.0))
+        {
+            m_z0 *= T(-1.0);
+            m_z  *= T(-1.0);
+        }
+
+        m_n0 = Vector<T,3>(T(0.0), m_z0, -m_y0);
+        m_n2 = Vector<T,3>(T(0.0), -m_z0, m_y1);
+        const Vector<T,3> n1 = Vector<T,3>(-m_z0, T(0.0), x1);
+        const Vector<T,3> n3 = Vector<T,3>(m_z0, T(0.0), -x0);
+
+        m_y0y0 = square(m_y0);
+        m_y1y1 = square(m_y1);
+        m_z0z0 = square(m_z0);
+
+        m_n0.z /= std::sqrt(m_z0z0 + m_y0y0);
+        m_n2.z /= std::sqrt(m_z0z0 + m_y1y1);
+        n1.z /= std::sqrt(m_z0z0 + square(x1));
+        n3.z /= std::sqrt(m_z0z0 + square(x0));
+
+        const T g0 = std::acos(-m_n0.z * n1.z);
+        const T g1 = std::acos(-n1.z * m_n2.z);
+        m_g2 = std::acos(-m_n2.z * n3.z);
+        m_g3 = std::acos(-n3.z * m_n0.z);
+
+        m_area = g0 + g1 + m_g2 + m_g3 - TwoPi<T>();
+    }
+
+    T get_area() const
+    {
+        return m_area;
+    }
+
+    Vector<T,3> sample(const Vector<T,2>& s) const
+    {
+        const T au = s[0] * m_area - m_g2 - m_g3 + TwoPi<T>();
```

Comment:
Does it makes sense to cache ` - m_g2 - m_g3 + TwoPi<T>();` ?  I'm wondering if the compiler does it by itself.

Reply:
We only compute this once, no?

Reply:
Because sample is called multiple times